### PR TITLE
feat: update FindPriorBackups to use backup index 

### PR DESCRIPTION
**Pull request by kev-cao**
This patch teaches `FindPriorBackups` (now more aptly renamed to `FindAllIncrementals`) to use the backup index introduced in #150384. Additional changes will still be required to fully fix the related tickets, namely updating the logic to no longer need to load ALL manifests of a chain into memory, only the manifests needed.

Epic: CRDB-47942

Informs: #150327, #150329, #150330, #150331

### CODE REVIEW

File: pkg/backup/backupinfo/manifest_handling.go [hunk 1]
```diff
@@ -1000,11 +1001,38 @@ func validateContinuity(manifests []backuppb.BackupManifest) error {
...
+// Note: This will mutate the underlying memory of the slices passed in as it
+// needs to sort them.
 func ElideSkippedLayers(
 	uris []string, backups []backuppb.BackupManifest, loc []jobspb.RestoreDetails_BackupLocalityInfo,
 ) ([]string, []backuppb.BackupManifest, []jobspb.RestoreDetails_BackupLocalityInfo, error) {
+	// First we sort our backups and their corresponding URIs/locality info by
+	// increasing end time and then by increasing start time to break ties. The
+	// latter ensures that when we elide duplicate end times, we keep the most
+	// compacted backup.
+	permutation := make([]int, len(backups))
```
Comment (jeffswenson) [inline pkg/backup/backupinfo/manifest_handling.go]: Permutation is neat, but there are simpler ways to implement `ElideSkippedLayers`. In general, I think its a bit of a code smell to pass around a bunch of slices that share an index. It has some uses in performance critical code, but in this case its easier to group related data into a shared struct. For example:

```go
type BackupChainEntry struct {
   uri string
   manifest backuppb.BackupManifest
   locality jobspb.RestoreDetails_backup_localityInfo
}

func CompareBackupChainEntry(a BackupChainEntry, b BackupChainEntry) int {
   // order logic here
}

func ElideSkippedLayers(backups []BackupChainEntry) ([]BackupChainEntry, error) {
   slices.SortFunc(backups, BackupChainEntry)
   ... rest of the logic
   return backups
}
```

This allows you to simplify the rest of the code ElideSkippedLayers as well. This allocates an extra slice, but slices.Delete is linear time, so its probably better to allocate then use Delete.

```
var result []BackupChainEntry
for i, backup := range sortedBackups {
   if i != 0 && sortedBackups[i-1].endTime() == backup.endTime() {
       continue
   }
   result = append(result, backup)
}
```

File: pkg/backup/backupdest/backup_destination_test.go [hunk 2]
```diff
@@ -351,47 +347,6 @@ func TestBackupRestoreResolveDestination(t *testing.T) {
 					// We also wrote a new full backup, so let's update the latest.
 					writeLatest(t, collectionLoc, expectedSuffix)
 				}
-
```
Comment (jeffswenson) [inline pkg/backup/backupdest/backup_destination_test.go]: Is there a test that covers the case that was removed here?

File: pkg/backup/backupdest/backup_index.go [hunk 3]
```diff
@@ -31,11 +35,25 @@ import (
 // information.
 func WriteBackupIndexMetadata(
 	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
 	user username.SQLUsername,
 	makeExternalStorageFromURI cloud.ExternalStorageFromURIFactory,
 	details jobspb.BackupDetails,
 ) error {
-	ctx, sp := tracing.ChildSpan(ctx, "backupdest.WriteBackupIndexMetadata")
+	indexStore, err := makeExternalStorageFromURI(
```
Comment (jeffswenson) [inline pkg/backup/backupdest/backup_index.go]: I think we are leaking this cloud storage instance and we need to defer a Close.

File: pkg/backup/backup_test.go [hunk 4]
```diff
@@ -718,7 +718,6 @@ func TestBackupAndRestoreJobDescription(t *testing.T) {
 	}
 
 	sqlDB.Exec(t, "BACKUP INTO ($1, $2, $3)", collections...)
-	sqlDB.Exec(t, "BACKUP INTO LATEST IN ($1, $2, $3)", collections...)
```
Comment (jeffswenson) [inline pkg/backup/backup_test.go]: Why was this removed from the test?

File: pkg/backup/backupdest/backup_destination.go [hunk 5]
```diff
@@ -237,41 +241,40 @@ func ResolveDest(
...
+		}
+		defer mem.Shrink(ctx, size)
+		startTime = precedingBackupManifest.EndTime
+		if startTime.IsEmpty() {
+			return ResolvedDestination{}, errors.Errorf("empty end time in prior backup manifest")
 		}
-		partName = partName + "-" + startTime.GoTime().Format(backupbase.DateBasedIncFolderNameSuffix)
 	}
+
+	partName := ConstructDateBasedIncrementalFolderName(startTime.GoTime(), endTime.GoTime())
```
Comment (jeffswenson) [inline pkg/backup/backupdest/backup_destination.go]: Is it safe to use the new partName format even in the case where we are not writing to the backup index?

**Review by jeffswenson — COMMENTED** [review]
Looks good overall 🥳

File: pkg/backup/backupinfo/manifest_handling.go (hunk 1 above)
Comment (kev-cao) [reply pkg/backup/backupinfo/manifest_handling.go]: Oh I like that approach a lot, I'll give it a shot. Thanks!

File: pkg/backup/backupdest/backup_destination_test.go (hunk 2 above)
Comment (kev-cao) [reply pkg/backup/backupdest/backup_destination_test.go]: This particular case is for where incrementals were previously written into the same directory as the full backup.

With these changes, this is no longer supported, so the test case was just removed instead of replaced.

File: pkg/backup/backupdest/backup_index.go (hunk 3 above)
Comment (kev-cao) [reply pkg/backup/backupdest/backup_index.go]: Oh yup we are totally leaking it here

File: pkg/backup/backup_test.go (hunk 4 above)
Comment (kev-cao) [reply pkg/backup/backup_test.go]: I removed it mainly for simplicity's sake. The test just checks the description outputted once the backup statement is fully resolved, so in this case, `INTO LATEST` just checks that `LATEST` is resolved accordingly.

We do already have a `BACKUP INTO LATEST ... WITH incremental_location` case which covers `LATEST`. If we wanted to still have them tested separately, it just would require a decent amount of changes to the test since you'd have to also update the logic for finding the full path that `LATEST` gets resolved into and separate the two full paths.

File: pkg/backup/backupdest/backup_destination.go (hunk 5 above)
Comment (kev-cao) [reply pkg/backup/backupdest/backup_destination.go]: Yup this is the current behavior. The new suffixed incremental path was introduced in 25.2. If I'm reading our restore policy on older backups correctly, 25.4 will support restores from 25.2+, so we can assume that all backups we read/write will contain the suffixed incremental path.

File: pkg/backup/backupdest/incrementals.go [hunk 6]
```diff
@@ -68,13 +72,93 @@ func CollectionsAndSubdir(paths []string, subdir string) ([]string, string, erro
...
+	defer sp.Finish()
+
+	indexes, err := ListIndexes(ctx, rootStore, subdir)
+	if err != nil {
+		return nil, err
+	}
+	// Due to our policy for writing indexes, if there exists an index file for a
+	// backup chain, we can assume the index is complete. So either an index has
+	// been written for every backup in the chain, or there are no indexes at all.
+	if len(indexes) == 0 {
```
Comment (msbutler) [inline pkg/backup/backupdest/incrementals.go:107]: since this is a big new feature, i think we should gate the _read_ path with a cluster setting, something like `read_with_index`, that is default false, and metamorphically set in tests. We will always write the index in 25.4+, but i think to minimize risk, we should keep the legacy read path available for now. We can make a separate decision later about flipping the default of this setting.

File: pkg/backup/backup_test.go [hunk 7]
```diff
@@ -9958,220 +9958,6 @@ func TestUserfileNormalizationIncrementalShowBackup(t *testing.T) {
 	sqlDB.Exec(t, query)
 }
 
-// TestBackupRestoreOldIncrementalDefaults tests that a call to restore
```
Comment (msbutler) [inline pkg/backup/backup_test.go]: to leave a better paper trail, could we spin off a separate commit that disables this legacy behavior?

File: pkg/backup/backupdest/backup_index.go [hunk 8]
```diff
@@ -138,6 +139,84 @@ func IndexExists(ctx context.Context, store cloud.ExternalStorage, subdir string
...
+// basename of its index file and its subdirectory. For full backups, the
+// returned path is relative to the root of the backup. For incremental
+// backups, the path is relative to the incremental storage location starting
+// from the subdir (e.g. 20250730/130000.00-20250730-120000.00). It expects
+// that subdir is resolved and not `LATEST`.
+//
+// Note: While the path is stored in the index file, we can take a shortcut here
+// and derive it from the filename solely because backup paths are
+// millisecond-precise and so are the timestamps encoded in the filename.
+func parseBackupFilePathFromIndexFileName(subdir, basename string) (string, error) {
```
Comment (msbutler) [inline pkg/backup/backupdest/backup_index.go:286]: I thought we were going to store the path to the full in the index file, so why do we need to parse the index file path?

File: pkg/backup/backupdest/backup_destination.go [hunk 9]
```diff
@@ -237,41 +237,40 @@ func ResolveDest(
 	}
 	prevBackupURIs = append([]string{plannedBackupDefaultURI}, prevBackupURIs...)
 
-	// Within the chosenSuffix dir, differentiate incremental backups with partName.
-	partName := endTime.GoTime().Format(backupbase.DateBasedIncFolderName)
-	if execCfg.Settings.Version.IsActive(ctx, clusterversion.V25_2) {
```
Comment (msbutler) [inline pkg/backup/backupdest/backup_destination.go]: nit: looks like you're removing a version gate in this part of the commit. I think this should be a separate commit.

**Review by msbutler — COMMENTED** [review]
flushing comments from first commit, as they may lead to some significant refactoring in the second commit :D. Also, I think this PR could be rebased on master (to remove the `TestBackupRestoreSeparateExplicitIsDefault ` diff)

File: pkg/backup/backupdest/backup_index.go (hunk 8 above)
Comment (kev-cao) [reply pkg/backup/backupdest/backup_index.go:286]: I was thinking that even though we do store the path in the index file, since we could actually parse it from the filename, we may as well do so to save us some reads.

And the full path would still be stored there as our source of truth in case we do ever mess with the filenames.

Comment (msbutler) [reply pkg/backup/backupdest/backup_index.go:286]: > we may as well do so to save us some reads

why do you think the cost of reading the index file will be expensive? IMHO the "cost" of maintaining this parsing logic significantly outweighs the "cost" loading the index file (which is tiny!)

**Comment by kev-cao** [comment]
@msbutler This is the PR that removes `TestBackupRestoreSeparateExplicitIsDefault` — the other PR removed writing to the same directory as the full, so a different situation.

File: pkg/backup/backupdest/backup_index.go (hunk 8 above)
Comment (kev-cao) [reply pkg/backup/backupdest/backup_index.go:286]: I suppose this isn't the bottleneck in the operations that use it 🤔 My consideration was for setting up the TCP connection and then doing the read, although I suppose the clients all use connection pools under the hood

Comment (msbutler) [reply pkg/backup/backupdest/backup_index.go:286]: restore/backup will be slow because they gotta read and write a whole cluster to external storage, not because they have to read a little metadata file ;)

Comment (kev-cao) [reply pkg/backup/backupdest/backup_index.go:286]: Yeah I'm falling for that premature optimization here — I was thinking of trying to keep it performant at the helper function level in the event that some other area of the codebase would use it. But since no other place actually does use it, I'm probably getting ahead of myself here.

File: pkg/backup/backupdest/incrementals.go [hunk 10]
```diff
@@ -70,13 +75,110 @@ func CollectionsAndSubdir(paths []string, subdir string) ([]string, string, erro
...
+			rel, err := filepath.Rel(
+				path.Join(backupbase.DefaultIncrementalsSubdir, subdir),
+				index.Path,
+			)
+			if err != nil {
+				return "", errors.Wrapf(err, "getting relative path for incremental backup")
+			}
+			// The old FindPriorBackups logic returned all paths with a leading path,
+			// so we do the same here for compatibility.
+			return string(backuputils.URLSeparator) + rel, nil
```
Comment (kev-cao) [inline pkg/backup/backupdest/incrementals.go]: Whether or not there is a leading/trailing slash is the new off-by-one error and I hate it.

File: pkg/backup/backuputils/utils.go [hunk 11]
```diff
@@ -102,13 +112,15 @@ func EncodeDescendingTS(ts time.Time) string {
...
+// will never start with a slash and will only end with a slash if the provied
+// backup URI ends in a slash.
 //
 // Example:
 //
 //	collectionURI: "nodelocal://1/collection"
-//	backupURI: "nodelocal://1/collection/backup1/"
-//	returns: "backup1/"
+//	backupURI: "nodelocal://1/collection/subdir/inc/"
+//	returns: "subdir/inc/"
```
Comment (kev-cao) [inline pkg/backup/backuputils/utils.go]: TODO: In a follow up PR, write tests to assert these sorts of guarantees.

File: pkg/backup/backupdest/incrementals.go [hunk 12]
```diff
@@ -70,13 +71,96 @@ func CollectionsAndSubdir(paths []string, subdir string) ([]string, string, erro
...
+	// 1. No incrementals have been taken yet, in which case the ListObjects calls
+	// will quickly terminate and there's little performance impact.
+	// 2. Incrementals have been taken, but they are in a custom non-default
+	// incremental location, in which case the legacy fallback would have been
+	// needed anyway.
+	if len(paths) == 0 {
+		return LegacyFindPriorBackups(ctx, incStore, includeManifest)
+	}
+
+	// Instead of passing in an argument that indicates whether a custom
```
Comment (msbutler) [inline pkg/backup/backupdest/incrementals.go]: hrm, theres like a 100 lines of code/comments that explain how we can check for a custom inc location. This puts a lot of mental overhead on the reader. I checked every caller of FindAllIncrementalPaths, and each can easily pass a bool on whether the caller is using incremental location. This feels a lot simpler.

File: pkg/backup/backupdest/incrementals.go [hunk 13]
```diff
@@ -121,19 +205,25 @@ func FindPriorBackups(
 // backupsFromLocation is a small helper function to retrieve all prior
 // backups from the specified location.
 func backupsFromLocation(
-	ctx context.Context, user username.SQLUsername, execCfg *sql.ExecutorConfig, loc string,
```
Comment (msbutler) [inline pkg/backup/backupdest/incrementals.go:124]: nit: it feels cleaner if you pass the defaultcollectionURI instead of rootStore. Then the helper will close the store instead of in ResolveIncrementalsBackupLocaiton.

File: pkg/backup/backupdest/backup_destination.go [hunk 14]
```diff
@@ -537,10 +544,13 @@ func ListFullBackupsInCollection(
 // included in the result, otherwise they are filtered out.
 func ResolveBackupManifests(
 	ctx context.Context,
+	execCfg *sql.ExecutorConfig,
```
Comment (msbutler) [inline pkg/backup/backupdest/backup_destination.go:548]: this function signature is a bit ridiculous. once the big PRs land, we should clean this up.  As an example, perhaps no stores should get passed into this monster. 

I also realize deleting the custom inc location code will also help a ton.

File: pkg/backup/compaction_job.go [hunk 15]
```diff
@@ -721,8 +721,18 @@ func getBackupChain(
...
+		return nil, nil, nil, nil, err
+	}
+	resolvedSubdir, err := resolveBackupSubdir(
+		ctx, execCfg, user, defaultCollectionURI, dest.Subdir,
+	)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
 	resolvedBaseDirs, resolvedIncDirs, _, err := resolveBackupDirs(
-		ctx, execCfg, user, dest.To, dest.IncrementalStorage, dest.Subdir,
```
Comment (msbutler) [inline pkg/backup/compaction_job.go:725]: what was wrong with dest.Subdir?

Comment (msbutler) [reply pkg/backup/compaction_job.go:725]: i'm quite confused. it looks like resolveBackupSubdir is also called within resolveBackupDIrs

File: pkg/backup/restore_planning.go [hunk 16]
```diff
@@ -1686,9 +1686,14 @@ func doRestorePlan(
...
 
+	defaultCollectionURI, _, err := backupdest.GetURIsByLocalityKV(from, "")
+	if err != nil {
+		return err
+	}
+
 	if strings.EqualFold(subdir, backupbase.LatestFileName) {
 		// set subdir to content of latest file
-		latest, err := backupdest.ReadLatestFile(ctx, from[0],
+		latest, err := backupdest.ReadLatestFile(ctx, defaultCollectionURI,
```
Comment (msbutler) [inline pkg/backup/restore_planning.go:1696]: what's this change for? Could we simply pass from[0] to resolveBackupManifests?

File: pkg/backup/show.go [hunk 17]
```diff
@@ -336,7 +336,11 @@ you must pass the 'encryption_info_dir' parameter that points to the directory o
 			info        backupInfo
 			memReserved int64
 		)
-		info.collectionURI = dest[0]
```
Comment (msbutler) [inline pkg/backup/show.go:339]: what's the reason for this change? could we simply pass dest[0] to ResolveBackupManifests?

File: pkg/backup/backupdest/incrementals.go (hunk 12 above)
Comment (kev-cao) [reply pkg/backup/backupdest/incrementals.go]: Yeah that sounds good to me. I don't think it necessarily applies here, but what do you think of this [coding guideline](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/181371303/Go+Golang+coding+guidelines#Try-to-avoid-bool-parameters) in our docs?

File: pkg/backup/backupdest/backup_destination.go (hunk 14 above)
Comment (kev-cao) [reply pkg/backup/backupdest/backup_destination.go:548]: Yeah it is absolutely horrifying to look at, especially if you're being onboarded onto this. I remember looking at this a year ago and having an aneurysm

File: pkg/backup/compaction_job.go (hunk 15 above)
Comment (kev-cao) [reply pkg/backup/compaction_job.go:725]: It was just because the default URI and a resolved subdir were both parameters that were added to `ResolveBackupManifests`, and since I already resolved the subdir outside of `resolveBackupDirs`, I figured I may as well pass it a resolved subdir to avoid another resolution step.

File: pkg/backup/restore_planning.go (hunk 16 above)
Comment (kev-cao) [reply pkg/backup/restore_planning.go:1696]: Sort of a similar situation with the last comment. I needed to resolve the default collection URI to pass into `ResolveBackupManifests`, so I just used that instead of `from[0]`. Actually, something I did notice was that in certain places we assume that `from[0]` is always the default URI but it's not a guarantee. So that means that if the user were to pass in a default locality but it wasn't the first URI in the list, then we would treat the wrong URI as the default. Of course, there's multiple places where these need fixed, but I'm thinking maybe we do something similar to subdir normalization. We could at the SQL layer in the statement planning ensure that the details always have the first URI be the default.

File: pkg/backup/show.go (hunk 17 above)
Comment (kev-cao) [reply pkg/backup/show.go:339]: Same reason as above

File: pkg/backup/backupdest/incrementals.go [hunk 18]
```diff
@@ -70,13 +71,73 @@ func CollectionsAndSubdir(paths []string, subdir string) ([]string, string, erro
...
+	rootStore cloud.ExternalStorage,
+	subdir string,
+	includeManifest bool,
+	customIncLocation bool,
+) ([]string, error) {
+	ctx, sp := tracing.ChildSpan(ctx, "backupdest.FindAllIncrementalPaths")
+	defer sp.Finish()
+
+	// Backup indexes do not support custom incremental locations.
+	if customIncLocation {
```
Comment (msbutler) [inline pkg/backup/backupdest/incrementals.go]: nit: these two bools could be combined

File: pkg/backup/backupdest/backup_index.go [hunk 19]
```diff
@@ -29,6 +30,15 @@ import (
 	"github.com/cockroachdb/errors"
 )
 
+var (
+	ReadBackupIndexEnabled = settings.RegisterBoolSetting(
+		settings.ApplicationLevel,
+		"backup.index.read.enabled",
+		"if true, the backup index will be read when reading from a backup collection",
+		false,
```
Comment (msbutler) [inline pkg/backup/backupdest/backup_index.go]: set it true metamorphically so CI will catch the bugs surfaced by e2e tests?

File: pkg/backup/backupdest/incrementals.go (hunk 12 above)
Comment (msbutler) [reply pkg/backup/backupdest/incrementals.go]: hrm, i think the guideline applies to other contexts, not this one. As you can tell here, adding the bool made the code much easier to read. 

Ss the guideline describes, I think a _bunch_ of bool args could be bad when they could instead be represented as _one_ enum type; or, if functional args would be better for a configuration struct-- that dave cheney blog post linked in the guideline is a banger.

File: pkg/backup/compaction_job.go (hunk 15 above)
Comment (msbutler) [reply pkg/backup/compaction_job.go:725]: To clarify, the argument to `resolveBackupDirs` should be `resolvedSubdir` not `subdir`, and within `resolveBackupDirs` and you can remove the `resolveBackupSubdir` call, right?

File: pkg/backup/restore_planning.go (hunk 16 above)
Comment (msbutler) [reply pkg/backup/restore_planning.go:1696]: > We could at the SQL layer in the statement planning ensure that the details always have the first URI be the default.

Huh, you're right that we dont enforce this, and yet in our [docs.](https://www.cockroachlabs.com/docs/stable/take-and-restore-locality-aware-backups#create-a-locality-aware-backup)...

>Note that the first URI in the list has to be the URI specified as the default URI when the backup was created. If you have moved your backups to a different location since the backup was originally taken, the first URI must be the new location of the files originally written to the default location.

I tried swapping the default uri around in a unit test and the darn test still passes. I have no idea how any of this has ever worked. I'll look into this after your prs land.

File: pkg/backup/backupdest/incrementals_test.go [hunk 20]
```diff
@@ -124,10 +127,366 @@ func TestJoinURLPath(t *testing.T) {
...
+	)
+	defer backupStore.Close()
+	require.NoError(t, err)
+	require.NoError(
+		t,
+		cloud.WriteFile(context.Background(), backupStore, backupbase.BackupMetadataName, emptyReader),
+	)
+	require.NoError(
+		t,
+		cloud.WriteFile(context.Background(), backupStore, backupbase.BackupManifestName, emptyReader),
```
Comment (msbutler) [inline pkg/backup/backupdest/incrementals_test.go:461]: nit: looks like you're writing the backup manifest twice

File: pkg/backup/backupdest/incrementals_test.go [hunk 21]
```diff
@@ -124,10 +127,366 @@ func TestJoinURLPath(t *testing.T) {
...
+			require.NoError(t, err)
+			incStore, err := execCfg.DistSQLSrv.ExternalStorageFromURI(
+				ctx, incStoreURI, username.RootUserName(),
+			)
+			require.NoError(t, err)
+
+			for _, chain := range tc.backupChains {
+				fullEnd := toTime(chain[0].end)
+
+				for idx, b := range chain {
```
Comment (msbutler) [inline pkg/backup/backupdest/incrementals_test.go]: if we shuffle the chain order, could we remove `TestListIndexes`. in other words, now that we have this test, what does TestListIndexes add?. I fear these PRs are adding _a lot_ of lines of test code. So, if we want to go back and tweak something, it will be hard to know what test to tweak.  We can also come back to this during cleanup.

File: pkg/backup/backupdest/incrementals.go [hunk 22]
```diff
@@ -191,23 +275,25 @@ func ResolveIncrementalsBackupLocation(
...
-
 	resolvedIncrementalsBackupLocation, err := backuputils.AppendPaths(fullBackupCollections, backupbase.DefaultIncrementalsSubdir, subdir)
 	if err != nil {
 		return nil, err
 	}
-
-	_, err = backupsFromLocation(ctx, user, execCfg, resolvedIncrementalsBackupLocation[0])
+	_, err = backupsFromLocation(
+		ctx, user, execCfg, defaultCollectionURI, subdir,
+		resolvedIncrementalsBackupLocation[0], false, /* customIncLocation */
```
Comment (msbutler) [inline pkg/backup/backupdest/incrementals.go]: nit, i'd instead define `customIncLocation := len(explicitIncrementalCollections) > 0 ` above the branch and pass that instead of hard coded bools. That protects against a future bug where someone monkeys around with the flow control here.

**Review by msbutler — COMMENTED** [review]
nearly there!

File: pkg/backup/backupdest/incrementals_test.go (hunk 20 above)
Comment (kev-cao) [reply pkg/backup/backupdest/incrementals_test.go:461]: Actually they write a metadata and a manifest file, needed since some of our logic still depends on one over the other :(

File: pkg/backup/backupdest/backup_index.go (hunk 19 above)
Comment (kev-cao) [reply pkg/backup/backupdest/backup_index.go]: Hmm, is that something we want to do at the registration level? That would mean that the value of the setting by default for a customer would be non-deterministic. I was thinking of setting it to `true` by default though

File: pkg/backup/compaction_job.go (hunk 15 above)
Comment (kev-cao) [reply pkg/backup/compaction_job.go:725]: Oh yeah I guess since `resolveBackupDirs` is only ever called here, then yeah we could put the onus on the caller to resolve subdirs.

File: pkg/backup/backupdest/backup_index.go (hunk 19 above)
Comment (msbutler) [reply pkg/backup/backupdest/backup_index.go]: if you wrapped the value like so:
```
metamorphic.ConstantWithTestBool("backup.index.read.enabled", false)
```

then our unit tests and non-perf roachtests will randomly use the index. for non-test clusters, this setting will continue to be default false.

File: pkg/backup/backupdest/incrementals_test.go (hunk 21 above)
Comment (kev-cao) [reply pkg/backup/backupdest/incrementals_test.go]: Yeah I don't mind deleting `TestListIndexes` — although I wonder what our testing philosophy here is. My principle was that `TestListIndexes` should be tested since it is its own independent helper. And therefore, when testing anything that relies on it, you don't have to worry about testing `ListIndexes` since you can assume that is already tested. Like in this particular test, shuffling the writes isn't testing an edge case of `FindAllIncrementalPaths`, it's testing an edge case of `ListIndexes`.

> So, if we want to go back and tweak something, it will be hard to know what test to tweak.

If `ListIndexes` were changed or some dependency breaks, then `TestListIndexes` fails as well as `TestFindAllIncrementalPaths`. And since `TestListIndexes` fails, we know that that's where the problem is and it narrows it down to that specific helper function.

And if we tweak something that relies on `ListIndexes`, the `TestListIndexes` function shouldn't need updating because the behavior of `ListIndexes` won't have changed.

---

That being said, I do agree that the current `TestListIndexes` isn't a great test so I'd be fine deleting it.

File: pkg/backup/backupdest/backup_index.go (hunk 19 above)
Comment (kev-cao) [reply pkg/backup/backupdest/backup_index.go]: Oooh TIL thanks!

File: pkg/backup/backupdest/incrementals_test.go (hunk 21 above)
Comment (msbutler) [reply pkg/backup/backupdest/incrementals_test.go]: lets chat more about this in person.

File: pkg/backup/backupdest/incrementals_test.go (hunk 20 above)
Comment (msbutler) [reply pkg/backup/backupdest/incrementals_test.go:461]: 💀

**Review by msbutler — APPROVED** [review]
One more big one!

**Comment by kev-cao** [comment]
TFTR!

bors r=msbutler
